src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs

```diff
@@ -0,0 +1,107 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public partial interface IOperation
+    {
+        public readonly partial struct ChildCollection
+        {
+            /// <summary>
+            /// Implements a reverse-order struct-based collection of <see cref="Operation"/> nodes.
+            /// This collection is ordered, but random access into the collection is not provided.
+            /// This type is not hardened to <code>default(Reversed)</code>, and will null reference
+            /// in these cases.
+            /// </summary>
+            [NonDefaultable]
+            public readonly struct Reversed : IReadOnlyCollection<IOperation>
+            {
+                private readonly Operation _operation;
+
+                internal Reversed(Operation operation)
+                {
+                    _operation = operation;
+                }
+
+                public int Count => _operation.ChildOperationsCount;
+
+                public Enumerator GetEnumerator() => new Enumerator(_operation);
+
+                public ImmutableArray<IOperation> ToImmutableArray()
+                {
+                    Enumerator enumerator = GetEnumerator();
+                    switch (_operation)
+                    {
+                        case NoneOperation { Children: var children }:
+                            return children;
+                        case InvalidOperation { Children: var children }:
+                            return children;
+                        case var _ when !enumerator.MoveNext():
```

Comment:
>var _ when !enumerator.MoveNext() [](http://example.com/codeflow?start=29&length=33)

The `default` branch depends on the execution order of `case` and `when` clauses. Please consider checking `Count == 0` here instead.